stan/math/prim/meta/is_container.hpp

```diff
@@ -15,12 +15,13 @@
 namespace stan {
 
 /**
- * Deduces whether type is eigen matrix or standard vector.
+ * Deduces whether type is a dense eigen type or standard vector.
  * @tparam Container type to check
  */
 template <typename Container>
-using is_container = bool_constant<
-    math::disjunction<is_eigen<Container>, is_std_vector<Container>>::value>;
+using is_container
```

Comment:
Wouldn't we also expect sparse matrices to be detected as a container though?

Reply:
So I'm kind of confused. It seems like a lot of the time we are declaring things containers when they are "vectorizable" so for sparse matrix I was like, "Well these are not really vectorizable so I would not say they are a container". imo I kind of feel like we should just have `container` simply mean a container pointing to items containing dynamically allocated memory like `std::vector<std::vector<Eigen::Matrix>>`. Like there is a few types of memory patterns that I'm thinking of here

`std::vector<double/var> & Eigen::Matrix<double/var, R, C>` where generic memory can be mapped to access linearly
`std::vector<std::vector<double/var> & Eigen::Matrix<double/var, R, C>>` where generic memory the container points to can be mapped to linearly
`Eigen::SparseMatrix<double/var>` If compressed, this can be mapped to generic memory and accessed linearly.

Hmm, maybe this PR is the wrong path and I should be using the `apply_vector_unary` pattern? Because then I would do what you are saying in another comment above and just calling the vectorized function over the memory